src/storage/src/render/persist_sink.rs

```diff
@@ -1182,28 +1187,125 @@ where
                 }
 
                 let result = {
-                    let _permit = busy_signal.acquire().await;
-                    write.compare_and_append_batch(
-                        &mut to_append[..],
-                        batch_lower.clone(),
-                        batch_upper.clone(),
-                    )
-                    .await
-                    .expect("Invalid usage")
+                    let maybe_err = if *read_only_rx.borrow() {
+
+                        // We have to wait for either us coming out of read-only
+                        // mode or someone else applying a write that covers our
+                        // batch.
+                        //
+                        // If we didn't wait for the latter here, and just go
+                        // around the loop again, we might miss a moment where
+                        // _we_ have to write down a batch. For example when our
+                        // input frontier advances to a state where we can
+                        // write, and the read-write instance sees the same
+                        // update but then crashes before it can append a batch.
+
+                        let maybe_err = loop {
+                            if collection_id.is_user() {
+                                tracing::debug!(
+                                    %worker_id,
+                                    %collection_id,
+                                    %shard_id,
+                                    ?batch_lower,
+                                    ?batch_upper,
+                                    ?current_upper,
+                                    "persist_sink is in read-only mode, waiting until we come out of it or the shard upper advances"
+                                );
+                            }
+
+                            // We don't try to be smart here, and for example
+                            // use `wait_for_upper_past()` and awaiting the
+                            // read-only flag. We'd have to use a select!, which
+                            // would require cancel safety of
+                            // `wait_for_upper_past()`, which it doesn't
+                            // advertise.
+                            tokio::time::sleep(Duration::from_secs(1)).await;
```

Comment:
if you want, there is a small optimization you can do here in case the read-only flag gets flipped right after entering the wait:

```rust
match tokio::time::timeout(Duration::from_secs(1), read_only_rx.changed()).await {
   ...
}
```